### ISSUE TITLE
Add streaming function to trees and forests 

### ISSUE THREAD

**Issue by PSSF23** [neurodata/treeple#40]
Current streaming trees and forests documentation: https://sdtf.neurodata.io/

@adam2392 Here is the fork streaming trees are relying on:

https://github.com/PSSF23/scikit-learn-stream

And here's the sklearn PR if you want to look at branch differences:

https://github.com/scikit-learn/scikit-learn/pull/18889

**Comment by adam2392** [comment]
@PSSF23 what are the specific Cython pieces in the fork that streaming trees relies on? Is it just the "update_node" function in `_tree.pyx/pxd`?

**Comment by PSSF23** [comment]
@adam2392 It's mainly the `update` function in `_tree.pyx`, and `_update_split_node`

**Comment by adam2392** [comment]
I see. Can you describe the update algorithm succinctly (e.g. what is it changing given an existing tree)? Maybe there is a way to just call the existing functions.

**Comment by PSSF23** [comment]
@adam2392 It takes the current batch of training data `(X,Y)`, sort them into leaf nodes by `tree.decision_path(X)`, and split the nodes (false roots) like starting new trees with those associated samples. I copied some code from `build`.

I have pseudocode in the `arxiv` draft as well: https://arxiv.org/pdf/2110.08483.pdf

**Comment by adam2392** [comment]
Thanks for linking. A few questions:

1. Whenever a tree is further split, is it only split once, or multiple times? 
2. Is there a case where the existing tree is not modified? E.g. whenever you stream new data, are all trees modified forsure?
3. Is it possible that different criterion and splitters are used across different streams of data? Or is it assumed that the incoming data must be split with the same splitter and measured with the same criterion?

Right now, it would be ideal if `build` is not copied, since it just makes it harder to maintain. What it seems like you want to do is "apply the tree to the samples" to get the leaf nodes, and then with remaining samples, continue the "building the tree process"? I think then it might be possible to do something like:

```Cython
cpdef update(...):
     # organize samples by decision_tree_path
     X, y, sample_weight = organize_tree(X, y, sample_weight)     

     # get leaf nodes
     leaf_nodes = tree.apply(X)
     
    # call build again
     self.build(tree, X, y, sample_weight)
```

which greatly simplifies the `update` function. However, this will require some exploration to see if it is possible cuz I'm not 100% sure. If you would like, you can open up a PR into the `tree-featuresv2` branch of the forked scikit-learn and we can iterate on adding this in. If it's not as simple, maybe we can refactor the `build` function so the initialization process can be overridden. This would probably also help your sklearn PR because there is no way they will accept a PR if it requires copy/pasting a lot of the existing build code.

By adding this ability to the forked version, all the models in scikit-tree should have "streaming capability" easily added via the Python API.

**Comment by PSSF23** [comment]
1. the child nodes of the leaves split will be recursively added as splitting candidates, so I had a stack in the `update` function that stores all "false roots." The splitting process is intended to be the same as batch decision tree.
2. the splitting criteria is user-defined like minimum number of samples at the leaf, so it's not tree-specific but node-specific. 
3. splitter and criterion could be technically modified between each update. I didn't do related experiments though.

I always use all the samples and associate them with leaf nodes they fall into. Then it's like building a new batch tree with associated samples at each leaf (thus "false roots"). I created a new `update` function mainly to avoid disrupting `build`, so batch learning still works as before on my fork.

If what you suggests works it will be awesome. I actually already have the changes in the fork on an outdated branch: https://github.com/neurodata/scikit-learn/tree/stream

**Comment by PSSF23** [comment]
I'm following the `partial_fit` api and resolved warnings and tests, before the PR became outdated. I stopped updating with sklearn `main` branch because the current version works, so I didn't want to disrupt that during experiments. We can definitely restart iterating.

**Comment by adam2392** [comment]
Yeah we should follow the partial fit API at the Python level.

Yeah if you're interested, I can review your PR to add the streaming func to the fork sklearn we have. Maintaining changes wrt upstream isn't that bad.

**Comment by PSSF23** [comment]
Appreciate the help! Here is the PR I made: https://github.com/neurodata/scikit-learn/pull/35

**Comment by adam2392** [comment]
This is addressed by #116 via @PSSF23
